cdap-watchdog/src/main/java/io/cdap/cdap/logging/clean/LogCleaner.java

```diff
@@ -62,6 +65,7 @@ public void run() {
         } else {
           deleteCount++;
           LOG.trace("File {} deleted by log cleanup", deletedEntry.getPath());
+          deleteParentFoldersIfEmpty(deletedEntry.getPath());
```

Comment:
To avoid calling location.list() unnecessarily, you can collect the set of parent locations that has child removed and do one check per parent directory in a separate loop. Something like this:

```
Set<Location> parents = new HashSet<>();
for (FileMetadataCleaner.DeletedEntry deletedEntry : deleteEntries) {
   Location locToDelete = Locations.getLocationFromAbsolutePath(locationFactory, deletedEntry.getPath());
   if (locToDelete.delete()) {
     parents.add(Locations.getParent(locToDelete);
   }
}

for (Location parent : parents) {
  deleteIfEmpty(parent);
}
```

Reply:
Removed the Location.list() call.

Location.delete() anyways returns false if the directory is not empty. If false is returned, break out of the loop.